include/wx/localedefs.h

```diff
@@ -108,6 +108,83 @@ struct WXDLLIMPEXP_BASE wxLanguageInfo
     const char* TrySetLocale() const;
 };
 
+// ----------------------------------------------------------------------------
+// wxLocaleIdent: allows to fully identify a locale under all platforms
+// ----------------------------------------------------------------------------
+
+class WXDLLIMPEXP_BASE wxLocaleIdent
+{
+public:
+    // Leave language empty
+    wxLocaleIdent() { }
+
+    // Construct from language, i.e. a two-letter ISO 639-1 code (or a
+    // three-letter ISO 639-2 code if there is no ISO 639-1 code for this
+    // language).
+    wxLocaleIdent(const char* language)
```

Comment:
I think — and tried to argue for that in the previous wxUILocale discussion — it is a mistake to construct language/locale identifier in this way in a modern API. Let's use the de-facto standard these days, which is [BCP 47 language tags](https://en.wikipedia.org/wiki/IETF_language_tag) ([brief intro](https://www.w3.org/International/articles/language-tags/)).

By all means, have a class useful for *parsing* or converting language identifier, but the primary input should be the tag: `wxLocaleIdent("es-005")`, not `wxLocaleIdent("es").Region("005")`. As the API is now, it doesn't allow construction of the identifier from a portable form (e.g. BCP-47 in a string, obtained from some API to determine user preference) and would require parsing in user code for that. And it already doesn't support full BCP-47 (dialects, private subtags), even without accounting for future expansions of the tags.

Reply:
This is a (strict) subset of BCP 47 language tags. Using full tags is problematic because how are we supposed to use them under Unix? If we parse them ourselves, we'd still be limited to the same subset, wouldn't we? And even macOS uses `language_REGION` format according to Apple docs and not BCP 47 `language-REGION` (although maybe it accepts both?).

I don't like passing the full tag to the ctor because then the current API doesn't make sense at all, e.g. `wxLocaleIdent("es-005").Region("ES")` is inconsistent. We could add `wxLocaleIdent::FromBCP47(string)` static ctor, but this is longer and I don't know how useful is it going to be. The current API was the best readability/safety/functionality compromise I could find.

FWIW I also thought about (additionally) providing a ctor from `wxLanguage` because even though our language database is incomplete, at least you can't make typos in language constants, unlike in strings.

To summarize, if we provide a ctor from BCP 47 string, we need to parse it ourselves, at least under Unix, and we couldn't provide safer/more readable components access as right now. Do we really want to do this?

Reply:
> And even macOS uses `language_REGION` format according to Apple docs and not BCP 47 `language-REGION` (although maybe it accepts both?).

macOS differentiates between Language and Locale IDs

https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html#//apple_ref/doc/uid/10000171i-CH15-SW1

Reply:
Yes, I should have mentioned that I read this and this is where I saw

> [language designator]-[script designator]_[region designator]

form for the locale identifiers. And we do need a locale, rather than language, identifier here.

Reply:
> Using full tags is problematic because how are we supposed to use them under Unix?

Parse the information you can use from them, ignore (but preserve) the rest.

> I don't like passing the full tag to the ctor because then the current API doesn't make sense at all,

That much I can agree with, but my conclusion from that is different: get rid of the .Region() etc. stuff. *That* makes little sense to me. Locales are not a thing you routinely hardcode like this.

> e.g. wxLocaleIdent("es-005").Region("ES") is inconsistent.

Is `wxLocaleIdent("es").Region("005").Region("ES")` consistent?

> We could add wxLocaleIdent::FromBCP47(string) static ctor, but this is longer and I don't know how useful is it going to be. 

I'd say more than wxLocaleIdent(language). I think this is part of my own confusion w.r.t. language vs locale in this API that I demonstrated. Language only is rarely a sufficient *locale* identification for UI-visible aspects (e.g. date formatting differs between fr-FR and fr-CA).

> FWIW I also thought about (additionally) providing a ctor from wxLanguage because even though our language database is incomplete, at least you can't make typos in language constants, unlike in strings.

I have no problem with that, or the argument for having it, but would like to point out the inconsistency between  `wxLocaleIdent(wxLANGUAGE_FRENCH_CANADIAN)` being fine, but `wxLocaleIdent("fr-CA")` not.

> To summarize, if we provide a ctor from BCP 47 string, we need to parse it ourselves, at least under Unix, and we couldn't provide safer/more readable components access as right now. Do we really want to do this?

We could provide *read* components access (by parsing). I don't think *write* access makes sense, but I'm probably missing something. 

More importantly, I think it is natural to read `wxLocaleIdent("foo")` (or just `"foo"` because the ctor is implicit) as meaning `"foo"` is the locale. But it isn't in this API, it is only the language, and `wxLocaleIdent("de-AT")` fails. This is unintuitive to me and I think it will lead to errors. Note that from this point of view, `wxLocaleIdent().Language("fr")` is fine. 

At the very least, the ctor should assert on presence of `-` or `_` in the language.